query-engine/connectors/mongodb-query-connector/src/logger.rs

```diff
@@ -0,0 +1,255 @@
+use crate::query_builder::MongoReadQuery;
+use itertools::Itertools;
+use mongodb::{
+    bson::{Bson, Document},
+    options::FindOptions,
+};
+use std::fmt::Write;
+use tracing::debug;
+
+// Bare-bones logging impl for debugging.
+pub(crate) fn log_read_query(coll_name: &str, query: &MongoReadQuery) {
+    let mut buffer = String::new();
+    fmt_query(&mut buffer, coll_name, query).unwrap();
+    debug!(target: "mongodb_query_connector::query", query = buffer.as_str(), item_type = "query", is_query = true);
+}
+
+macro_rules! write_indented {
+    ($buffer:expr, $depth:expr, $fmt_str:literal, $($args:expr)*) => {
+        write!($buffer, "{}{}", indent($depth), format!($fmt_str, $($args)*))?;
+    };
+}
+
+fn fmt_query(buffer: &mut String, coll_name: &str, query: &MongoReadQuery) -> std::fmt::Result {
+    match query {
+        MongoReadQuery::Find(find) => {
+            write!(buffer, "db.{}.find(", coll_name)?;
+
+            if let Some(ref filter) = find.filter {
+                fmt_doc(buffer, filter, 1)?;
+                write!(buffer, ", ")?;
+            }
+
+            fmt_opts(buffer, &find.options, 1)?;
+            write!(buffer, ")")
+        }
+        MongoReadQuery::Pipeline(pipeline) => {
+            write!(buffer, "db.{}.aggregate(", coll_name)?;
+
+            let stages: Vec<_> = pipeline
+                .stages
+                .iter()
+                .map(|stage| Bson::Document(stage.clone()))
+                .collect();
+
+            fmt_list(buffer, &stages, 1)?;
+            write!(buffer, ")")
+        }
+    }
+}
+
+fn fmt_opts(buffer: &mut String, opts: &FindOptions, depth: usize) -> std::fmt::Result {
+    if cfg!(debug_assertions) {
+        writeln!(buffer, "{{")?;
+    } else {
+        write!(buffer, "{{")?;
+    }
+
+    if let Some(skip) = opts.skip {
+        write_indented!(buffer, depth, "skip: {},\n", skip);
+    }
+
+    if let Some(limit) = opts.limit {
+        write_indented!(buffer, depth, "limit: {},\n", limit);
+    }
+
+    if let Some(ref sort) = opts.sort {
+        write_indented!(buffer, depth, "sort: ",);
+        fmt_doc(buffer, sort, depth + 1)?;
+
+        if cfg!(debug_assertions) {
+            writeln!(buffer, ",")?;
+        } else {
+            write!(buffer, ",")?;
+        }
+    }
+
+    if let Some(ref projection) = opts.projection {
+        write_indented!(buffer, depth, "projection: ",);
+        fmt_doc(buffer, projection, depth + 1)?;
+
+        if cfg!(debug_assertions) {
+            writeln!(buffer)?;
+        }
+    }
+
+    write!(buffer, "}}")
+}
+
+#[cfg(debug_assertions)]
+fn indent(depth: usize) -> String {
+    " ".repeat(4 * depth)
+}
+
+#[cfg(not(debug_assertions))]
+fn indent(_: usize) -> String {
+    String::from(" ")
+}
+
+fn fmt_doc(buffer: &mut String, doc: &Document, depth: usize) -> std::fmt::Result {
+    if cfg!(debug_assertions) {
+        writeln!(buffer, "{{")?;
+    } else {
+        write!(buffer, "{{")?;
+    }
+
```

Comment:
This repeats a lot, can probably be a separate function / macro

Reply:
I kind of try, but it is not just _that_ pattern but the whole stuff with indentation. I will end up creating a lot of very small macros and functions so I avoided it and probably the whole stuff could be just a giant macro (not really giant but "selective")